### PR TITLE
fix: Order of callbacks invocations, if TxnOp.Merge is used 

### PR DESCRIPTION
**Pull request by michaljurecko**
Jira: https://keboola.atlassian.net/browse/PSGO-596

All tests will pass in the final [PR](https://github.com/keboola/keboola-as-code/pull/1770).

[Large set of fixes and E2e tests](https://github.com/keboola/keboola-as-code/pull/1759) is divided into smaller PRs to do a review.

--------------

**Changes:**
- Fixed order of  callback invocations, if `TxnOp.Merge` is used.
- Aded method `TxnOp.FirstErrorOnly`.

-----------

Commend per commits.

-----------

image

---------

At first, please read `TxnOp` docs:

https://github.com/keboola/keboola-as-code/blob/1d5dffdda7b8375ab05800820be1b8860168b497/internal/pkg/service/common/etcdop/op/txn.go#L19-L34

### CODE REVIEW

File: internal/pkg/service/common/etcdop/op/txn.go [hunk 1]
```diff
@@ -26,29 +34,32 @@ import (
...
-	processors  []func(ctx context.Context, r *TxnResult[R])
-	ifs         []etcd.Cmp
-	thenOps     []etcd.Op
-	elseOps     []etcd.Op
-	thenMappers []MapFn
-	elseMappers []MapFn
+	result     *R
+	client     etcd.KV
+	processors []func(ctx context.Context, r *TxnResult[R])
+	opCounter  map[txnPartType]int
```
Comment (hosekpeter) [inline internal/pkg/service/common/etcdop/op/txn.go:59]: Why did you add this?

**Review by hosekpeter — APPROVED** [review]

File: internal/pkg/service/common/etcdop/op/txn.go (hunk 1 above)
Comment (Matovidlo) [reply internal/pkg/service/common/etcdop/op/txn.go:59]: looks like telemetry stuff

**Review by Matovidlo — APPROVED** [review]

File: internal/pkg/service/common/etcdop/op/txn.go [hunk 2]
```diff
@@ -199,92 +219,120 @@ func (v *TxnOp[R]) Op(ctx context.Context) (LowLevelOp, error) {
...
+
+func (v *lowLevelTxn[R]) addPart(ctx context.Context, part txnPart[R]) error {
+	opIndex := v.opCounter[part.Type]
+	v.opCounter[part.Type]++
+
+	// Add if
+	if part.Type == txnOpIf {
+		v.ifs = append(v.ifs, part.If)
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/common/etcdop/op/txn.go:282]: :heavy_check_mark:  I assume it somehow doesn't need the low-level operation 

Yes, it is `IF` condition, not `THEN/ELSE` operation, so there is (always) `factory=nil`.

File: internal/pkg/service/common/etcdop/op/txn.go (hunk 1 above)
Comment (michaljurecko) [reply internal/pkg/service/common/etcdop/op/txn.go:59]: no, no telemetry, ... but better errors messages, let see where it is used:

https://github.com/keboola/keboola-as-code/blob/ea0baab5900336e5496428f2192060c154a572be/internal/pkg/service/common/etcdop/op/txn.go#L275

https://github.com/keboola/keboola-as-code/blob/ea0baab5900336e5496428f2192060c154a572be/internal/pkg/service/common/etcdop/op/txn.go#L292

https://github.com/keboola/keboola-as-code/blob/ea0baab5900336e5496428f2192060c154a572be/internal/pkg/service/common/etcdop/op/txn.go#L297

File: internal/pkg/service/common/etcdop/op/txn.go [hunk 3]
```diff
@@ -8,6 +8,14 @@ import (
 	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
 )
 
+const (
+	txnOpIf      = txnPartType("if")
+	txnOpThen    = txnPartType("then")
+	txnOpThenTxn = txnPartType("thenTxn") // thenTxnOps are separated from thenOps to avoid confusing between Then and Merge
```
[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/common/etcdop/op/txn.go:14]: You use `Merge` almost all the time, e.g. create File AND Slice, so we merge CreateFile and CreateSlice transactions to a new one.

There will be many examples in other PRs, you can also find some occurrences yourself in: https://github.com/keboola/keboola-as-code/pull/1759

File: internal/pkg/service/common/etcdop/op/txn.go (hunk 1 above)
Comment (Matovidlo) [reply internal/pkg/service/common/etcdop/op/txn.go:59]: Could we choose different name? I expected that Counter is something tight with telemetry. Wouldn't it be better to have `operations int` ? Looks as number of operations within the transaction. No counter name needed :)

Comment (michaljurecko) [reply internal/pkg/service/common/etcdop/op/txn.go]: I  imroved it a bit, see https://github.com/keboola/keboola-as-code/pull/1771/commits/418425119c8ea23a0a7eccc60d406920b358f8f2

@hosekpeter @Matovidlo please be a little more precise in your review, you could definitely find out why it is there, it is used in one place, thanks :beers: 

I suggest to open IDE with the [final code](https://github.com/keboola/keboola-as-code/pull/1759), to see full picture.

File: internal/pkg/service/common/etcdop/op/txn.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (michaljurecko) [reply internal/pkg/service/common/etcdop/op/txn.go:14]: Additional note: For non-txn operation, `Merge` and `Then` are equal, both add the op to the `then` branch.

File: internal/pkg/service/common/etcdop/op/txn.go (hunk 1 above)
Comment (Matovidlo) [reply internal/pkg/service/common/etcdop/op/txn.go]: Looks alright.

**Review by Matovidlo — APPROVED** [review]
LGTM.

**Review by hosekpeter — APPROVED** [review]
LGTM
